File: Packages/StrandAnalytics/Sources/StrandAnalytics/EffectRanker.swift
import Foundation

// EffectRanker.swift — the unified, LAG-AWARE "what moves your Charge" ranker.
//
// Pure, deterministic, DB-free. Generalises ActivityCostEngine's single-sport D+1 to
// EVERY logged journal behaviour against EVERY daily outcome, and searches a small fixed
// lag set so it can tell "same day" from "shows up the next morning."
//
// For each (behaviour b, outcome o, lag L ∈ {0, +1, +2}):
//   1. behaviourDays = the day keys b was logged on (dose ≥ 1; callers pass the set).
//   2. Pair behaviour day D with outcome day D+L by SHIFTING the outcome map back by L —
//      i.e. re-key outcome[D+L] under D via CorrelationEngine.shiftDay(day, by: -L). This is
//      exactly the alignment ActivityCostEngine does for D+1, parameterised over the lag and
//      reusing the same fixed-UTC day arithmetic, so behaviour day D is compared with the
//      outcome that landed L days later.
//   3. Run the EXISTING BehaviorInsights.effect on the shifted outcome map → meanWith,
//      meanWithout, delta, cohensD, Welch pApprox, significant (already gated at
//      min(nWith, nWithout) ≥ 5).
//   4. Keep, per (b, o), the lag L* with the LARGEST |cohensD| among lags whose effect
//      computed AND cleared the significance group gate, carrying L* as the lead/lag.
//
// HONESTY (effect-size first, not stargazing): the primary signal is the effect SIZE +
// n + a ScoreConfidence tier, never a bare "significant" stamp. The lag search is capped
// at the small fixed set {0,1,2} so the comparison count stays bounded and explainable; we
// never claim a behaviour "causes" anything — only that it lines up with a change.
//
// Output: one RankedEffect per (b, o) that produced a usable lag, ranked with the same rule
// as BehaviorInsights.rank (significant first, |cohensD| desc, stable tiebreak), so the feed
// matches the existing Behaviour Effects ordering exactly. Self-contained except for the two
// reused primitives, so the Kotlin twin is line-for-line.
//
// (Spec: 2026-06-19-v5-insights-correlation-engine-design.md — "Lag-aware effect ranking".)

// MARK: - Result

/// One ranked, lag-aware behaviour→outcome effect: the best lag's BehaviorEffect plus the
/// lead/lag it was found at and a confidence tier from the paired-day count.
public struct RankedEffect: Equatable, Sendable {
    /// The behaviour label (e.g. "Alcohol").
    public let behavior: String
    /// The outcome metric label (e.g. "Charge").
    public let outcome: String
    /// The lag (in days) at which the strongest honest effect was found: 0 = same day,
    /// +1 = the next morning, +2 = two mornings later.
    public let lag: Int
    /// The measured effect at `lag` (means, delta, Cohen's d, Welch p, significant).
    public let effect: BehaviorEffect
    /// Per-result certainty tier from the smaller group's size at the chosen lag.
    public let confidence: ScoreConfidence

    public init(behavior: String, outcome: String, lag: Int,
                effect: BehaviorEffect, confidence: ScoreConfidence) {
        self.behavior = behavior
        self.outcome = outcome
        self.lag = lag
        self.effect = effect
        self.confidence = confidence
    }

    /// Plain-English lead/lag chip text, e.g. "same day" / "next morning" / "2 mornings later".
    public var leadLagText: String {
        switch lag {
        case 0: return "same day"
        case 1: return "next morning"
        default: return "\(lag) mornings later"
        }
    }

    /// The sign-aware sentence for this row, reusing BehaviorInsights' renderer plus the
    /// lead/lag clause so each card reads "…, showing up the next morning."
    public func sentence() -> String {
        let base = BehaviorInsights.sentence(effect)
        // Drop the trailing period, append the lead/lag, restore it.
        let trimmed = base.hasSuffix(".") ? String(base.dropLast()) : base
        return "\(trimmed) (\(leadLagText))."
    }
}

// MARK: - Engine

public enum EffectRanker {

    /// The fixed, bounded lag set searched per (behaviour, outcome). Kept small and explicit
    /// — not a fitted VAR — so the multiple-comparison count stays bounded and honest at small n.
    public static let lagSet: [Int] = [0, 1, 2]

    /// Paired-day count below which a chosen lag's confidence is `.calibrating` (too thin to
    /// shout). This equals BehaviorInsights' significance group gate, so a row that clears the
    /// gate is never `.calibrating`.
    public static let calibratingBelow: Int = BehaviorInsights.minGroupForSignificance  // 5
    /// Paired-day count at/above which a chosen lag's confidence is `.solid` (else `.building`).
    public static let solidPairs: Int = 10

    // MARK: - Rank

    /// Rank every behaviour against one outcome across the lag set, keeping each behaviour's
    /// best lag. Mirrors BehaviorInsights.rank's ordering on the surviving rows.
    ///
    /// - Parameters:
    ///   - behaviors: per behaviour name, the SET of "yyyy-MM-dd" days it was logged (dose ≥ 1).
    ///   - outcomeByDay: the daily outcome series keyed "yyyy-MM-dd" (e.g. Charge 0–100).
    ///   - outcome: the outcome label carried onto each RankedEffect (e.g. "Charge").
    /// - Returns: one RankedEffect per behaviour that produced a usable lag, ranked
    ///   significant-first, |cohensD| desc, then behaviour name asc. Behaviours with no
    ///   computable lag are dropped.
    public static func rank(behaviors: [String: Set<String>],
                            outcomeByDay: [String: Double],
                            outcome: String) -> [RankedEffect] {
        var rows: [RankedEffect] = []
        // Sort behaviour names so the build order is deterministic regardless of dict order.
        for name in behaviors.keys.sorted() {
            let days = behaviors[name]!
            if let row = bestLag(behaviorDays: days, outcomeByDay: outcomeByDay,
                                 behavior: name, outcome: outcome) {
                rows.append(row)
            }
        }
        return sorted(rows)
    }

    /// Find the best-lag RankedEffect for ONE behaviour against ONE outcome, or nil when no
    /// lag in `lagSet` yields a computable effect that clears the group gate.
    public static func bestLag(behaviorDays: Set<String>,
                               outcomeByDay: [String: Double],
                               behavior: String,
                               outcome: String) -> RankedEffect? {
        var best: (lag: Int, effect: BehaviorEffect)?
        for lag in lagSet {
            let shifted = shiftedOutcome(outcomeByDay, byLag: lag)
            guard let e = BehaviorInsights.effect(behaviorDays: behaviorDays,
                                                  outcomeByDay: shifted,
                                                  behavior: behavior,
                                                  outcome: outcome) else { continue }
            // Group gate: a lag only competes if both sides clear the significance minimum,
            // so a 2-day lag can't win on a fluke. (Mirrors BehaviorInsights.significant's n-gate.)
            guard Swift.min(e.nWith, e.nWithout) >= BehaviorInsights.minGroupForSignificance else { continue }

            if let cur = best {
                // Largest |cohensD| wins; ties break to the SMALLER lag (prefer same-day /
                // shorter lead over a longer one when the effect size is identical).
                let better = abs(e.cohensD) > abs(cur.effect.cohensD)
                    || (abs(e.cohensD) == abs(cur.effect.cohensD) && lag < cur.lag)
                if better { best = (lag, e) }
            } else {
                best = (lag, e)
            }
        }
        guard let chosen = best else { return nil }
        let pairs = Swift.min(chosen.effect.nWith, chosen.effect.nWithout)
        return RankedEffect(behavior: behavior, outcome: outcome, lag: chosen.lag,
                            effect: chosen.effect, confidence: confidence(forPairs: pairs))
    }

    // MARK: - Lag alignment

    /// Re-key the outcome series so behaviour day D is paired with the outcome that landed
    /// `lag` days later: out'[D] = out[D+lag]. We move the VALUE from key D+lag back to key D
    /// by shifting each existing key BACKWARD by `lag` (shiftDay(day, by: -lag)). Then a plain
    /// BehaviorInsights split on behaviour day D reads the D+lag outcome — the same join
    /// ActivityCostEngine performs for D+1, generalised. `lag == 0` is the identity map.
    static func shiftedOutcome(_ outcomeByDay: [String: Double], byLag lag: Int) -> [String: Double] {
        if lag == 0 { return outcomeByDay }
        var out: [String: Double] = [:]
        out.reserveCapacity(outcomeByDay.count)
        for (day, value) in outcomeByDay {
            // The outcome ON day `day` belongs to behaviour day `day - lag`.
            if let behaviourKey = CorrelationEngine.shiftDay(day, by: -lag) {
                out[behaviourKey] = value
            }
        }
        return out
    }

    // MARK: - Confidence

    /// Confidence from the smaller group's paired-day count: below the group gate →
    /// `.calibrating`; gate…<solidPairs → `.building`; ≥ solidPairs → `.solid`.
    static func confidence(forPairs pairs: Int) -> ScoreConfidence {
        if pairs < calibratingBelow { return .calibrating }
        return pairs >= solidPairs ? .solid : .building
    }

    // MARK: - Ranking

    /// Stable rank matching BehaviorInsights.rank: significant first, |cohensD| desc, then
    /// behaviour name ascending.
    static func sorted(_ rows: [RankedEffect]) -> [RankedEffect] {
        rows.sorted { a, b in
            if a.effect.significant != b.effect.significant { return a.effect.significant }
            let la = abs(a.effect.cohensD), lb = abs(b.effect.cohensD)
            if la != lb { return la > lb }
            return a.behavior < b.behavior
        }
    }
}

File: android/app/src/main/java/com/noop/analytics/EffectRanker.kt
package com.noop.analytics

import kotlin.math.abs
import kotlin.math.exp
import kotlin.math.sqrt

/*
 * EffectRanker.kt — the unified, LAG-AWARE "what moves your Charge" ranker.
 *
 * Faithful Kotlin mirror of StrandAnalytics/EffectRanker.swift. Keep the lag set, the
 * shift-by-lag alignment, the Welch t / Cohen's d effect math, the best-lag selection, the
 * confidence ladder, and the ranking byte-identical to Swift — cross-platform parity is the
 * contract.
 *
 * Pure, deterministic, DB-free. Generalises ActivityCostEngine's single-sport D+1 to EVERY
 * logged journal behaviour against EVERY daily outcome, searching a small fixed lag set
 * {0,+1,+2} so it can tell "same day" from "shows up the next morning."
 *
 * For each (behaviour b, outcome o, lag L):
 *   1. behaviourDays = the day keys b was logged on (dose ≥ 1; callers pass the set).
 *   2. Pair behaviour day D with outcome day D+L by SHIFTING the outcome map back by L —
 *      re-key outcome[D+L] under D via shiftDay(day, -L). (Same arithmetic ActivityCostEngine
 *      uses for D+1, parameterised over the lag.)
 *   3. Compute the effect (means, delta, Cohen's d via pooled SD, Welch p via a normal-CDF
 *      tail) — the byte-identical port of Swift BehaviorInsights.effect — gated at
 *      min(nWith, nWithout) ≥ 5.
 *   4. Keep, per (b, o), the lag with the LARGEST |cohensD| among lags that cleared the gate.
 *
 * HONESTY (effect-size first): the primary signal is the effect SIZE + n + a ScoreConfidence
 * tier, never a bare "significant" stamp; the lag set is capped so the comparison count stays
 * bounded; copy never claims a behaviour "causes" anything.
 *
 * (Spec: 2026-06-19-v5-insights-correlation-engine-design.md — "Lag-aware effect ranking".)
 */

/**
 * The measured effect of one behaviour on one outcome metric at a single lag. Byte-identical
 * port of Swift's BehaviorEffect (the shared correlation substrate has no standalone Kotlin
 * BehaviorInsights, so the struct + math live here).
 */
data class BehaviorEffect(
    val behavior: String,
    val outcome: String,
    /** Mean outcome on days the behaviour WAS logged. */
    val meanWith: Double,
    /** Mean outcome on days the behaviour was NOT logged. */
    val meanWithout: Double,
    /** meanWith − meanWithout (signed). */
    val delta: Double,
    /** Percent change of meanWith relative to meanWithout, or null when meanWithout is 0. */
    val pctChange: Double?,
    /** Number of behaviour-present days with an outcome value. */
    val nWith: Int,
    /** Number of behaviour-absent days with an outcome value. */
    val nWithout: Int,
    /** Cohen's d using the pooled SD (signed; sign matches delta). */
    val cohensD: Double,
    /** Two-sided p-value (Welch t-test, normal approximation). */
    val pApprox: Double,
    /** pApprox < 0.05 AND min(nWith, nWithout) ≥ 5. */
    val significant: Boolean,
)

/**
 * One ranked, lag-aware behaviour→outcome effect: the best lag's BehaviorEffect plus the
 * lead/lag it was found at and a confidence tier from the paired-day count.
 */
data class RankedEffect(
    val behavior: String,
    val outcome: String,
    /** 0 = same day, +1 = the next morning, +2 = two mornings later. */
    val lag: Int,
    val effect: BehaviorEffect,
    val confidence: ScoreConfidence,
) {
    /** Plain-English lead/lag chip text. */
    val leadLagText: String
        get() = when (lag) {
            0 -> "same day"
            1 -> "next morning"
            else -> "$lag mornings later"
        }

    /** The sign-aware sentence plus the lead/lag clause. */
    fun sentence(): String {
        val base = EffectRanker.sentence(effect)
        val trimmed = if (base.endsWith(".")) base.dropLast(1) else base
        return "$trimmed ($leadLagText)."
    }
}

object EffectRanker {

    /** The fixed, bounded lag set searched per (behaviour, outcome). */
    val lagSet: List<Int> = listOf(0, 1, 2)

    /** Minimum group size (each side) for an effect to be flagged significant. Mirrors Swift
     *  BehaviorInsights.minGroupForSignificance. */
    const val minGroupForSignificance: Int = 5

    /** Significance threshold on the approximate p-value. */
    const val alpha: Double = 0.05

    /** Paired-day count below which a chosen lag's confidence is CALIBRATING. */
    const val calibratingBelow: Int = minGroupForSignificance

    /** Paired-day count at/above which a chosen lag's confidence is SOLID (else BUILDING). */
    const val solidPairs: Int = 10

    // Rank.

    /**
     * Rank every behaviour against one outcome across the lag set, keeping each behaviour's
     * best lag. Mirrors Swift's ordering on the surviving rows.
     *
     * @param behaviors per behaviour name, the SET of "yyyy-MM-dd" days it was logged (dose ≥ 1).
     * @param outcomeByDay the daily outcome series keyed "yyyy-MM-dd".
     * @param outcome the outcome label carried onto each RankedEffect.
     */
    fun rank(
        behaviors: Map<String, Set<String>>,
        outcomeByDay: Map<String, Double>,
        outcome: String,
    ): List<RankedEffect> {
        val rows = ArrayList<RankedEffect>()
        for (name in behaviors.keys.sorted()) {
            val days = behaviors.getValue(name)
            val r = bestLag(days, outcomeByDay, name, outcome)
            if (r != null) rows.add(r)
        }
        return sorted(rows)
    }

    /** Find the best-lag RankedEffect for ONE behaviour against ONE outcome, or null. */
    fun bestLag(
        behaviorDays: Set<String>,
        outcomeByDay: Map<String, Double>,
        behavior: String,
        outcome: String,
    ): RankedEffect? {
        var bestLagValue = 0
        var bestEffect: BehaviorEffect? = null
        for (lag in lagSet) {
            val shifted = shiftedOutcome(outcomeByDay, lag)
            val e = effect(behaviorDays, shifted, behavior, outcome) ?: continue
            if (minOf(e.nWith, e.nWithout) < minGroupForSignificance) continue

            val cur = bestEffect
            if (cur == null) {
                bestEffect = e
                bestLagValue = lag
            } else {
                val better = abs(e.cohensD) > abs(cur.cohensD) ||
                    (abs(e.cohensD) == abs(cur.cohensD) && lag < bestLagValue)
                if (better) {
                    bestEffect = e
                    bestLagValue = lag
                }
            }
        }
        val chosen = bestEffect ?: return null
        val pairs = minOf(chosen.nWith, chosen.nWithout)
        return RankedEffect(behavior, outcome, bestLagValue, chosen, confidence(pairs))
    }

    // Lag alignment.

    /**
     * Re-key the outcome series so behaviour day D pairs with outcome[D+lag]: move the value ON
     * day `day` back to key `day − lag` via shiftDay(day, −lag). lag == 0 is the identity map.
     */
    internal fun shiftedOutcome(outcomeByDay: Map<String, Double>, lag: Int): Map<String, Double> {
        if (lag == 0) return outcomeByDay
        val out = HashMap<String, Double>(outcomeByDay.size)
        for ((day, value) in outcomeByDay) {
            val behaviourKey = ActivityCostEngine.shiftDay(day, -lag) ?: continue
            out[behaviourKey] = value
        }
        return out
    }

    // Effect (byte-identical port of Swift BehaviorInsights.effect + helpers).

    /**
     * Compute the effect of [behavior] on [outcome]. Days are partitioned into "with"
     * (day ∈ behaviorDays) and "without", restricted to days with an outcome value. Returns
     * null unless both groups are non-empty AND there are ≥ 3 points total.
     */
    internal fun effect(
        behaviorDays: Set<String>,
        outcomeByDay: Map<String, Double>,
        behavior: String,
        outcome: String,
    ): BehaviorEffect? {
        val withVals = ArrayList<Double>()
        val withoutVals = ArrayList<Double>()
        for ((day, value) in outcomeByDay) {
            if (behaviorDays.contains(day)) withVals.add(value) else withoutVals.add(value)
        }

        val n1 = withVals.size
        val n2 = withoutVals.size
        if (n1 < 1 || n2 < 1 || n1 + n2 < 3) return null

        val m1 = withVals.sum() / n1
        val m2 = withoutVals.sum() / n2
        val delta = m1 - m2

        val pct: Double? = if (m2 != 0.0) delta / abs(m2) * 100.0 else null

        val v1 = sampleVariance(withVals, m1)
        val v2 = sampleVariance(withoutVals, m2)

        val d = cohensD(m1, m2, n1, v1, n2, v2)
        val p = welchP(m1, v1, n1, m2, v2, n2)

        val sig = p < alpha && minOf(n1, n2) >= minGroupForSignificance

        return BehaviorEffect(
            behavior = behavior, outcome = outcome,
            meanWith = m1, meanWithout = m2, delta = delta, pctChange = pct,
            nWith = n1, nWithout = n2, cohensD = d, pApprox = p, significant = sig,
        )
    }

    /** Render an effect as a plain-English sentence (byte-identical to Swift BehaviorInsights). */
    internal fun sentence(e: BehaviorEffect): String {
        val directionWord = when {
            e.delta > 0 -> "higher"
            e.delta < 0 -> "lower"
            else -> "unchanged"
        }
        val magnitude = when {
            e.delta == 0.0 -> "no different"
            e.pctChange != null -> "${roundedInt(abs(e.pctChange))}% $directionWord"
            else -> "${round1(abs(e.delta))} $directionWord"
        }
        val avgWith = roundedInt(e.meanWith)
        val avgWithout = roundedInt(e.meanWithout)
        return "On days you logged ‘${e.behavior}’, ${e.outcome} was $magnitude " +
            "(avg $avgWith vs $avgWithout, n=${e.nWith} vs ${e.nWithout})."
    }

    // Confidence.

    /** Below the gate → CALIBRATING; gate…<solidPairs → BUILDING; ≥ solidPairs → SOLID. */
    internal fun confidence(pairs: Int): ScoreConfidence {
        if (pairs < calibratingBelow) return ScoreConfidence.CALIBRATING
        return if (pairs >= solidPairs) ScoreConfidence.SOLID else ScoreConfidence.BUILDING
    }

    // Ranking.

    /** Stable rank: significant first, |cohensD| desc, then behaviour name ascending. */
    internal fun sorted(rows: List<RankedEffect>): List<RankedEffect> =
        rows.sortedWith(
            compareByDescending<RankedEffect> { it.effect.significant }
                .thenByDescending { abs(it.effect.cohensD) }
                .thenBy { it.behavior },
        )

    // Statistics helpers (byte-identical to Swift).

    /** Sample variance (ddof = 1). 0 for fewer than 2 values. */
    internal fun sampleVariance(values: List<Double>, mean: Double): Double {
        val n = values.size
        if (n < 2) return 0.0
        var ss = 0.0
        for (v in values) {
            val d = v - mean
            ss += d * d
        }
        return ss / (n - 1)
    }

    /** Cohen's d with the pooled SD. 0 when df ≤ 0 or the pooled SD is 0. */
    internal fun cohensD(m1: Double, m2: Double, n1: Int, v1: Double, n2: Int, v2: Double): Double {
        val df = n1 + n2 - 2
        if (df <= 0) return 0.0
        val pooledVar = ((n1 - 1) * v1 + (n2 - 1) * v2) / df
        val sp = sqrt(pooledVar)
        if (sp <= 0) return 0.0
        return (m1 - m2) / sp
    }

    /** Two-sided Welch t-test p-value with a normal-approximation tail. */
    internal fun welchP(m1: Double, v1: Double, n1: Int, m2: Double, v2: Double, n2: Int): Double {
        val se2 = v1 / n1 + v2 / n2
        if (se2 <= 0) return if (m1 == m2) 1.0 else 0.0
        val t = (m1 - m2) / sqrt(se2)
        return 2.0 * (1.0 - normalCDF(abs(t)))
    }

    /** Standard-normal CDF Φ(z) using the A&S 7.1.26 erf approximation. */
    internal fun normalCDF(z: Double): Double = 0.5 * (1.0 + erfApprox(z / sqrt(2.0)))

    /** erf(x) — Abramowitz & Stegun 7.1.26, |error| ≤ 1.5e-7. */
    internal fun erfApprox(x: Double): Double {
        val sign = if (x < 0) -1.0 else 1.0
        val ax = abs(x)
        val t = 1.0 / (1.0 + 0.3275911 * ax)
        val y = 1.0 - (((((1.061405429 * t - 1.453152027) * t) + 1.421413741) * t -
            0.284496736) * t + 0.254829592) * t * exp(-ax * ax)
        return sign * y
    }

    // Formatting helpers.

    /** Round to nearest integer, HALF AWAY FROM ZERO — matches Swift's Int(x.rounded())
     *  (Kotlin's Math.round / roundToInt round half toward +∞, which would diverge on a
     *  negative .5; we replicate Swift's .toNearestOrAwayFromZero rule exactly). */
    internal fun roundedInt(x: Double): Int {
        val sign = if (x < 0) -1.0 else 1.0
        return (sign * Math.floor(abs(x) + 0.5)).toInt()
    }

    /** Round to one decimal place, half away from zero (mirrors Swift (x*10).rounded()/10). */
    internal fun round1(x: Double): Double {
        val scaled = x * 10.0
        val sign = if (scaled < 0) -1.0 else 1.0
        return sign * Math.floor(abs(scaled) + 0.5) / 10.0
    }
}
